crates/ruff/src/rules/flake8_bugbear/snapshots/ruff__rules__flake8_bugbear__tests__B006_B006_B008.py.snap

```diff
@@ -1,113 +1,351 @@
+87 87 | 
+88 88 | 
+89    |-def this_too(value=collections.OrderedDict()):
+   89 |+def this_too(value=None):
+   90 |+    if value is None:
+   91 |+        value = collections.OrderedDict()
+90 92 |     ...
+91 93 | 
+92 94 | 
+
+B006_B008.py:93:32: B006 [*] Do not use mutable data structures for argument defaults
    |
-92 | def dict_comprehension_also_not_okay(default={i: i**2 for i in range(3)}):
-   |                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ B006
-93 |     pass
+93 | async def async_this_too(value=collections.defaultdict()):
+   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^ B006
+94 |     ...
    |
+   = help: Replace mutable data structure with `None` in argument default and replace it with data structure inside the function if still `None`
 
-B006_B008.py:96:45: B006 Do not use mutable data structures for argument defaults
+ℹ Possible fix
+90 90 |     ...
+91 91 | 
+92 92 | 
+93    |-async def async_this_too(value=collections.defaultdict()):
+   93 |+async def async_this_too(value=None):
+   94 |+    if value is None:
+   95 |+        value = collections.defaultdict()
+94 96 |     ...
+95 97 | 
+96 98 | 
+
+B006_B008.py:97:26: B006 [*] Do not use mutable data structures for argument defaults
    |
-96 | def set_comprehension_also_not_okay(default={i**2 for i in range(3)}):
-   |                                             ^^^^^^^^^^^^^^^^^^^^^^^^ B006
-97 |     pass
+97 | def dont_forget_me(value=collections.deque()):
+   |                          ^^^^^^^^^^^^^^^^^^^ B006
+98 |     ...
    |
+   = help: Replace mutable data structure with `None` in argument default and replace it with data structure inside the function if still `None`
 
-B006_B008.py:100:33: B006 Do not use mutable data structures for argument defaults
+ℹ Possible fix
+94  94  |     ...
+95  95  | 
+96  96  | 
+97      |-def dont_forget_me(value=collections.deque()):
+    97  |+def dont_forget_me(value=None):
+    98  |+    if value is None:
+    99  |+        value = collections.deque()
+98  100 |     ...
+99  101 | 
+100 102 | 
+
+B006_B008.py:102:46: B006 [*] Do not use mutable data structures for argument defaults
+    |
+101 | # N.B. we're also flagging the function call in the comprehension
+102 | def list_comprehension_also_not_okay(default=[i**2 for i in range(3)]):
+    |                                              ^^^^^^^^^^^^^^^^^^^^^^^^ B006
+103 |     pass
     |
-100 | def kwonlyargs_mutable(*, value=[]):
+    = help: Replace mutable data structure with `None` in argument default and replace it with data structure inside the function if still `None`
+
+ℹ Possible fix
+99  99  | 
+100 100 | 
+101 101 | # N.B. we're also flagging the function call in the comprehension
+102     |-def list_comprehension_also_not_okay(default=[i**2 for i in range(3)]):
+    102 |+def list_comprehension_also_not_okay(default=None):
+    103 |+    if default is None:
+    104 |+        default = [(i ** 2) for i in range(3)]
```

Comment:
Any idea why these parenthesis are being added here as the default value doesn't contain them? Same thing is happening for other comprehensions as well.

Reply:
I assumed it was intended behaviour of `checker.generator().expr(default)` which for `ListComp`'s (and other comprehensions) unparses in two parts:

```rust
Expr::ListComp(ast::ExprListComp {
                elt,
                generators,
                range: _range,
            }) => {
                self.p("[");
                self.unparse_expr(elt, precedence::MAX);
                self.unparse_comp(generators);
                self.p("]");
            }
```
The `unparse_expr` func adds parentheses conditinoally here, before matching the `ast`:
```rust
macro_rules! group_if {
            ($lvl:expr, $body:block) => {{
                let group = level > $lvl;
                self.p_if(group, "(");
                let ret = $body;
                self.p_if(group, ")");
                ret
            }};
        }
```

I figured this added the parentheses and was intended but looking at the tests of the generator now (e.g.:

```rust
        assert_round_trip!("{i for i in b async for i in a if await i for b in i}");
```

I'm not entirely sure this is the case anymore

Reply:
Yeah, this seems like a bug:

```rust
assert_round_trip!("[i*2 for i in range(5)]");
// thread 'source_code::generator::tests::tmp_test' panicked at 'assertion failed: `(left == right)`
//  left: `"[(i * 2) for i in range(5)]"`,
// right: `"[i*2 for i in range(5)]"`', crates/ruff_python_ast/src/source_code/generator.rs:1491:9
// note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
// test source_code::generator::tests::tmp_test ... FAILED
```

/cc @MichaReiser

Reply:
Omg, too much macro use in our `Generator`. This is a precedence issue that's probably worth opening an issue for.

This code decides whether to parenthesize an expression or not 

https://github.com/astral-sh/ruff/blob/58a79c1392003ec63904fe20a33e7b62857c3db1/crates/ruff_python_ast/src/source_code/generator.rs#L874-L881

The `oprec` macro took me a while to understand but precedence either gets assigned the `AND` (33) or `OR` (31) precedence (the same is true for bin op, it's just that they use slightly different values for the precedence)

Now, the generator adds parentheses if the parent's precedence is higher. We use the `MAX` precedence for the list comprehension, which means all expressions get parenthesized

https://github.com/astral-sh/ruff/blob/58a79c1392003ec63904fe20a33e7b62857c3db1/crates/ruff_python_ast/src/source_code/generator.rs#L1010-L1019

The fix requires identifying the right precedence and passing it through. A second (unrelated) refactor could be to use an enum for the precedence instead of an `u8`.

Reply:
Tracking this in #5777